### PR TITLE
refactor: update arithmetic operations for checked ones

### PR DESCRIPTION
**Pull request by eguajardo**
## Description

https://axelarnetwork.atlassian.net/browse/AXE-3274

## Todos

- [ ] Unit tests
- [ ] Manual tests
- [ ] Documentation
- [ ] Connect epics/issues

### CODE REVIEW

File: contracts/rewards/src/contract/execute.rs [hunk 1]
```diff
@@ -192,12 +200,14 @@ where
         // (i.e. we are in epoch 0, which started at block 0 and epoch duration is 1000. At epoch 500, the params
         // are updated to shorten the epoch duration to 100 blocks. We set the epoch number to 1, to prevent skipping
         // epochs 1-4, and so all events prior to the start of epoch 1 have an epoch number of 0)
-        let should_end =
-            cur_epoch.block_height_started + u64::from(new_params.epoch_duration) < block_height;
+        let should_end = cur_epoch
+            .block_height_started
+            .saturating_add(u64::from(new_params.epoch_duration))
```
Comment (cgorenflo) [inline contracts/rewards/src/contract/execute.rs:205]: while the event is highly unlikely, if it happens the current epoch is suddenly shorter than what we expect based on the epoch duration. I don't think we should break that invariant, so I'd prefer a `checked_add()?` here

File: contracts/rewards/src/contract/execute.rs [hunk 2]
```diff
@@ -192,12 +200,14 @@ where
...
+            < block_height;
         let cur_epoch = if should_end {
             Epoch {
                 block_height_started: block_height,
-                epoch_num: cur_epoch.epoch_num + 1,
+                epoch_num: cur_epoch.epoch_num.saturating_add(1),
```
Comment (cgorenflo) [inline contracts/rewards/src/contract/execute.rs:210]: `checked_add().expect()`
the epoch number can never be larger than block_height

File: contracts/rewards/src/contract/execute.rs [hunk 3]
```diff
@@ -215,7 +225,11 @@ where
         amount: nonempty::Uint128,
     ) -> Result<(), ContractError> {
         let mut pool = self.store.load_rewards_pool(pool_id)?;
-        pool.balance += Uint128::from(amount);
+        pool.balance = pool
+            .balance
+            .checked_add(Uint128::from(amount))
+            .map_err(Into::<ContractError>::into)
```
Comment (cgorenflo) [inline contracts/rewards/src/contract/execute.rs:231]: does `into()` not work here?

File: contracts/rewards/src/contract/execute.rs [hunk 4]
```diff
@@ -231,12 +245,17 @@ where
...
     rewards_2
         .into_iter()
-        .fold(rewards_1, |mut rewards, (addr, amt)| {
-            *rewards.entry(addr).or_default() += amt;
-            rewards
+        .try_fold(rewards_1, |mut rewards, (addr, amt)| {
+            let r = rewards.entry(addr).or_default();
+            *r = r
+                .checked_add(amt)
+                .map_err(Into::<ContractError>::into)
```
Comment (cgorenflo) [inline contracts/rewards/src/contract/execute.rs:255]: does `into()` not work here?

File: contracts/rewards/src/contract/execute.rs [hunk 5]
```diff
@@ -231,12 +245,17 @@ where
...
+                .map_err(Report::from)?;
+
+            Ok(rewards)
```
Comment (cgorenflo) [inline contracts/rewards/src/contract/execute.rs:258]:
```suggestion
            Ok(r)
```

Comment (cgorenflo) [reply contracts/rewards/src/contract/execute.rs:258]: this not failing a unit test indicates that we are missing coverage. Can you ensure we properly check the rewards?

File: contracts/rewards/src/error.rs [hunk 6]
```diff
@@ -36,6 +37,9 @@ pub enum ContractError {
     #[error("specified block has already passed")]
     BlockHeightInPast,
 
+    #[error("{0}")]
```
Comment (cgorenflo) [inline contracts/rewards/src/error.rs:40]:
```suggestion
    #[error(transparent)]
```

File: contracts/rewards/src/state.rs [hunk 7]
```diff
@@ -200,7 +200,7 @@ impl RewardsPool {
         if self.balance < reward {
             return Err(ContractError::PoolBalanceInsufficient.into());
         }
-        self.balance -= reward;
+        self.balance = self.balance.saturating_sub(reward);
```
Comment (cgorenflo) [inline contracts/rewards/src/state.rs:203]:
```
self.balance = self.balance.checked_sub(reward).map_err(|_| ContractError::PoolBalanceInsufficient.into())
```

File: contracts/service-registry/src/state.rs [hunk 8]
```diff
@@ -89,7 +89,7 @@ impl BondingState {
             | BondingState::Unbonding {
                 amount,
                 unbonded_at: _,
-            } => amount + to_add,
+            } => amount.saturating_add(to_add),
```
Comment (cgorenflo) [inline contracts/service-registry/src/state.rs:92]: very dangerous, because we are taking one amount for staking that's being sent through the bank module and potentially set a different amount here. This must be `checked_add()?`

File: contracts/voting-verifier/src/execute.rs [hunk 9]
```diff
@@ -58,7 +58,7 @@ pub fn verify_worker_set(
                 source_chain: config.source_chain,
                 source_gateway_address: config.source_gateway_address,
                 confirmation_height: config.confirmation_height,
-                expires_at: env.block.height + config.block_expiry,
+                expires_at: env.block.height.saturating_add(config.block_expiry),
```
Comment (cgorenflo) [inline contracts/voting-verifier/src/execute.rs:61]: again, I'd rather use the checked version here because otherwise the expiry interval could have an unexpected length

File: contracts/voting-verifier/src/execute.rs [hunk 10]
```diff
@@ -145,7 +145,7 @@ pub fn verify_messages(
...
```
Comment (cgorenflo) [inline contracts/voting-verifier/src/execute.rs:148]: same here

File: contracts/voting-verifier/src/execute.rs [hunk 11]
```diff
@@ -259,7 +259,7 @@ fn create_worker_set_poll(
 ) -> Result<PollId, ContractError> {
     let id = POLL_ID.incr(store)?;
 
-    let poll = WeightedPoll::new(id, snapshot, block_height + expiry, 1);
+    let poll = WeightedPoll::new(id, snapshot, block_height.saturating_add(expiry), 1);
```
Comment (cgorenflo) [inline contracts/voting-verifier/src/execute.rs:262]: same here

File: contracts/voting-verifier/src/execute.rs [hunk 12]
```diff
@@ -274,7 +274,7 @@ fn create_messages_poll(
 ) -> Result<PollId, ContractError> {
     let id = POLL_ID.incr(store)?;
 
-    let poll = WeightedPoll::new(id, snapshot, block_height + expiry, poll_size);
+    let poll = WeightedPoll::new(id, snapshot, block_height.saturating_add(expiry), poll_size);
```
Comment (cgorenflo) [inline contracts/voting-verifier/src/execute.rs:277]: same here

File: packages/axelar-wasm-std/src/counter.rs [hunk 13]
```diff
@@ -23,7 +21,7 @@ impl<'a, T: AddAssign + Copy + Default + One + Serialize + DeserializeOwned> Cou
 
     pub fn incr(&self, store: &mut dyn Storage) -> StdResult<T> {
         let mut value = self.cur(store);
-        value += T::one();
+        value = value.saturating_add(&T::one());
```
Comment (cgorenflo) [inline packages/axelar-wasm-std/src/counter.rs:24]: since `incr` already returns a result, I think it makes more sense to use a checked_add here, so it doesn't behave in an unexpected manner

File: packages/axelar-wasm-std/src/threshold.rs [hunk 14]
```diff
@@ -127,7 +129,7 @@ impl TryFrom<Threshold> for MajorityThreshold {
     type Error = Error;
 
     fn try_from(value: Threshold) -> Result<Self, Error> {
-        if value.numerator() <= value.denominator() / Uint64::from(2u64) {
+        if value.numerator() <= value.denominator().div(Uint64::from(2u64)) {
```
Comment (cgorenflo) [inline packages/axelar-wasm-std/src/threshold.rs:132]: what's the reason for this change? Does this get around the clippy warning?

File: packages/axelar-wasm-std/src/voting.rs
Comment (cgorenflo) [inline packages/axelar-wasm-std/src/voting.rs:1]: why are arithmetic functions implemented on PollId in the first place? That doesn't make any sense

File: packages/axelar-wasm-std/src/voting.rs [hunk 15]
```diff
@@ -179,10 +188,10 @@ impl Tallies {
     }
 
     pub fn tally(&mut self, vote: &Vote, weight: &Uint256) {
-        *self
-            .0
-            .get_mut(&vote.to_string())
-            .unwrap_or(&mut Uint256::zero()) += weight;
+        let mut binding = Uint256::zero();
```
Comment (cgorenflo) [inline packages/axelar-wasm-std/src/voting.rs:191]:
```suggestion
        let mut zero = Uint256::zero();
```

**Review by cgorenflo — CHANGES_REQUESTED** [review]

File: ampd/src/broadcaster.rs [hunk 16]
```diff
@@ -165,14 +167,16 @@ where
             .change_context(Error::TxBuilding)?;
 
         self.estimate_gas(sim_tx).await.map(|gas| {
-            let gas_adj = (gas as f64 * self.config.gas_adjustment) as u64;
+            let gas_adj = gas as f64 * self.config.gas_adjustment;
 
             Fee::from_amount_and_gas(
                 Coin {
-                    amount: (gas_adj as f64 * self.config.gas_price.amount).ceil() as u128,
+                    amount: cast((gas_adj.mul(self.config.gas_price.amount)).ceil())
```
[the remark this answers is not in the document]
Comment (eguajardo) [reply ampd/src/broadcaster.rs:174]: yeah neither `into()` nor `try_into()` work. The `mul` was indeed to get around clippy lint, clippy doesn't really flags multiplication of two floats, because any "overflowing" value is set to `f64::INFINITY` which is a valid value. In this case it's flagged because `amount` is a custom struct even though it behaves like a float `pub struct FiniteAmount(f64)`, so I just basically delegated the lint checking to the `mul` implementation

File: contracts/multisig-prover/src/execute.rs [hunk 17]
```diff
@@ -283,10 +283,12 @@ pub fn should_update_worker_set(
...
         .count()
-        + cur_workers_signers
-            .iter()
-            .filter(|item| !new_workers_signers.contains(item))
-            .count()
+        .saturating_add(
+            cur_workers_signers
+                .iter()
+                .filter(|item| !new_workers_signers.contains(item))
+                .count(),
```
[the remark this answers is not in the document]
Comment (eguajardo) [reply contracts/multisig-prover/src/execute.rs]: I ended up doing something a bit different to avoid unnecessary insertion to an intermediary structure

File: contracts/rewards/src/contract/execute.rs (hunk 3 above)
Comment (eguajardo) [reply contracts/rewards/src/contract/execute.rs:231]: tried, but it doesn't work

File: contracts/rewards/src/contract/execute.rs (hunk 4 above)
Comment (eguajardo) [reply contracts/rewards/src/contract/execute.rs:255]: tried, but it doesn't work

File: contracts/rewards/src/contract/execute.rs (hunk 5 above)
Comment (eguajardo) [reply contracts/rewards/src/contract/execute.rs:258]: `r` it's a mutable reference to the rewards entry, so it mutates rewards. This is why tests don't fail. I have changed the code to make it less confusing

File: packages/axelar-wasm-std/src/voting.rs
Comment (eguajardo) [reply packages/axelar-wasm-std/src/voting.rs:1]: was wondering the same thing, but turns out it's a requirement for the `Counter` struct, it require `One` trait which in turn requires `Mul` trait, also it needs the addition trait to be able to increase the counter

File: ampd/src/queue/msg_queue.rs [hunk 18]
```diff
@@ -9,7 +9,10 @@ pub struct MsgQueue {
 impl MsgQueue {
     pub fn push(&mut self, msg: Any, gas_cost: Gas) {
         self.msgs.push(msg);
-        self.gas_cost += gas_cost;
+        self.gas_cost = self
+            .gas_cost
+            .checked_add(gas_cost)
+            .expect("arithmetic operation overflows");
     }
```
[the remark this answers is not in the document]
Comment (eguajardo) [reply ampd/src/queue/msg_queue.rs:16]: I changed it to throw an error instead of panic, however I don't really like changing the signature to consume and return self, I think it's counterintuitive since usually other structs with a function similar to `push()` like Vec and HashMap are expected to be mutable to add new elements. I also think it's not necessary since we are not chaining push to some other functions

File: contracts/multisig/src/contract/execute.rs [hunk 19]
```diff
@@ -29,12 +29,14 @@ pub fn start_signing_session(
...
-            counter += Uint64::one();
+            counter = counter
+                .checked_add(Uint64::one())
+                .map_err(StdError::overflow)?;
             Ok(counter)
         },
     )?;
 
-    let expires_at = env.block.height + config.block_expiry;
+    let expires_at = env.block.height.saturating_add(config.block_expiry);
```
[the remark this answers is not in the document]
Comment (eguajardo) [reply contracts/multisig/src/contract/execute.rs:39]: in general, when the overflow was not caused by user input from the current transaction, then I lean towards saturation, the reason is that I prefer to let the protocol continue with some value rather than bricking the protocol. In this particular case I think it's more likely that the overflow is caused due to a misconfigured block_expiry (or a bug in the network?) and if that's the case then new signing sessions won't be produced until something is fixed

File: contracts/rewards/src/state.rs (hunk 7 above)
Comment (eguajardo) [reply contracts/rewards/src/state.rs:203]: yeah, at first I was thinking that it doesn't really matter since there was a check before, but this actually can replace that previous check

File: contracts/voting-verifier/src/execute.rs (hunk 9 above)
Comment (eguajardo) [reply contracts/voting-verifier/src/execute.rs:62]: I kind of feel the same way as here: https://github.com/axelarnetwork/axelar-amplifier/pull/296#discussion_r1525694763

File: contracts/voting-verifier/src/execute.rs (hunk 10 above)
Comment (eguajardo) [reply contracts/voting-verifier/src/execute.rs:149]: I kind of feel the same way as here: https://github.com/axelarnetwork/axelar-amplifier/pull/296#discussion_r1525694763

File: contracts/voting-verifier/src/execute.rs (hunk 11 above)
Comment (eguajardo) [reply contracts/voting-verifier/src/execute.rs:263]: I kind of feel the same way as here: https://github.com/axelarnetwork/axelar-amplifier/pull/296#discussion_r1525694763

File: contracts/voting-verifier/src/execute.rs (hunk 12 above)
Comment (eguajardo) [reply contracts/voting-verifier/src/execute.rs:278]: I kind of feel the same way as here: https://github.com/axelarnetwork/axelar-amplifier/pull/296#discussion_r1525694763

File: packages/axelar-wasm-std/src/threshold.rs (hunk 14 above)
Comment (eguajardo) [reply packages/axelar-wasm-std/src/threshold.rs:132]: yes, I wanted to delegate the handling to the underlying implementation like I did with `FiniteAmount`, in this case is Uint64 which panics when there is an issue during the div, but now that I think about it, I think this is not a good idea with structs external to us like Uint64 because clippy won't be able to detect an issue if implementation changes

File: ampd/src/queue/msg_queue.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (cgorenflo) [reply ampd/src/queue/msg_queue.rs]: even with a fluent interface the struct is mutable, we wouldn't create an immutable copy. Checking the current code, we use loops over pushes in multiple places, the reason there is no chaining is that we currently can't chain because it's interface doesn't allow it. A fluent interface would also allow to use it with `fold()`-like functions

File: contracts/rewards/src/contract/execute.rs [hunk 20]
```diff
@@ -192,12 +205,24 @@ where
...
+            .block_height_started
+            .checked_add(u64::from(new_params.epoch_duration))
+            .ok_or_else(|| {
+                StdError::overflow(OverflowError::new(
+                    OverflowOperation::Add,
+                    cur_epoch.block_height_started,
+                    new_params.epoch_duration,
+                ))
+            })
+            .map_err(ContractError::from)?
```
Comment (cgorenflo) [inline contracts/rewards/src/contract/execute.rs:218]: why is this double conversion necessary when there is already a `ContractError::Overflow`?

File: contracts/rewards/src/error.rs [hunk 21]
```diff
@@ -1,8 +1,12 @@
 use axelar_wasm_std_derive::IntoContractError;
+use cosmwasm_std::{OverflowError, StdError};
 use thiserror::Error;
 
 #[derive(Error, Debug, PartialEq, IntoContractError)]
 pub enum ContractError {
+    #[error(transparent)]
+    Std(#[from] StdError),
```
Comment (cgorenflo) [inline contracts/rewards/src/error.rs:8]: if at all possible, I'd like to avoid a catchall Std error below the root level. This makes it too easy to default wrapping everything in this which contradicts the purpose of error-stack to add context to the errror

File: contracts/multisig/src/contract/execute.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (cgorenflo) [reply contracts/multisig/src/contract/execute.rs:39]: a wrong configuration will lead to potentially wrong votes from this ampd instance, so the verifier could get penalized and worst case thrown out without being aware of what the problem was

File: contracts/voting-verifier/src/execute.rs (hunk 9 above)
Comment (cgorenflo) [reply contracts/voting-verifier/src/execute.rs]: see my response there

File: contracts/voting-verifier/src/execute.rs (hunk 10 above)
Comment (cgorenflo) [reply contracts/voting-verifier/src/execute.rs]: see my response there

File: contracts/voting-verifier/src/execute.rs (hunk 11 above)
Comment (cgorenflo) [reply contracts/voting-verifier/src/execute.rs]: see my response there

File: contracts/voting-verifier/src/execute.rs (hunk 12 above)
Comment (cgorenflo) [reply contracts/voting-verifier/src/execute.rs]: see my response there

File: packages/axelar-wasm-std/src/voting.rs
Comment (cgorenflo) [reply packages/axelar-wasm-std/src/voting.rs:1]: omg, that sucks. Could you add comments on those trait implementations to explain that they are necessary to satisfy the `One` trait?

File: ampd/src/queue/msg_queue.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (fish-sammy) [reply ampd/src/queue/msg_queue.rs]: I am OK with `push` consuming self and therefore enabling chaining, but I am not sure what difference panic v.s. returning error here makes?

File: contracts/multisig-prover/src/execute.rs [hunk 22]
```diff
@@ -274,27 +277,13 @@ pub fn should_update_worker_set(
...
-            .filter(|item| !new_workers_signers.contains(item))
-            .count()
-        > max_diff
+    signers_difference_count(&new_workers.signers, &cur_workers.signers).saturating_add(
+        signers_difference_count(&cur_workers.signers, &new_workers.signers),
+    ) > max_diff
+}
+
+fn signers_difference_count(s1: &BTreeMap<String, Signer>, s2: &BTreeMap<String, Signer>) -> usize {
+    s1.values().filter(|v| !s2.values().contains(v)).count()
```
[the remark this answers is not in the document]
Comment (eguajardo) [reply contracts/multisig-prover/src/execute.rs:286]: what do you mean is symetric? it's not. https://en.wikipedia.org/wiki/Complement_(set_theory)#Relative_complement. Here is how is described for BTreeSet: https://doc.rust-lang.org/std/collections/struct.BTreeSet.html#method.difference

File: ampd/src/queue/msg_queue.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (eguajardo) [reply ampd/src/queue/msg_queue.rs]: could we leave the refactoring to a fluent interface struct out of scope of this PR?

File: contracts/multisig-prover/src/execute.rs (hunk 22 above)
[the remark this answers is not in the document]
Comment (cgorenflo) [reply contracts/multisig-prover/src/execute.rs:286]: you are right, my bad. I could've sworn that what is defined as symmetric difference is just called difference

File: contracts/multisig-prover/src/execute.rs [hunk 23]
```diff
@@ -274,27 +277,13 @@ pub fn should_update_worker_set(
...
-        .iter()
-        .filter(|item| !cur_workers_signers.contains(item))
-        .count()
-        + cur_workers_signers
-            .iter()
-            .filter(|item| !new_workers_signers.contains(item))
-            .count()
-        > max_diff
+    signers_difference_count(&new_workers.signers, &cur_workers.signers).saturating_add(
+        signers_difference_count(&cur_workers.signers, &new_workers.signers),
```
Comment (cgorenflo) [inline contracts/multisig-prover/src/execute.rs:281]: extract this into `signers_symmetric_difference_count` to make it easier to understand the condition

**Review by cgorenflo — APPROVED** [review]
